modules/runner/src/integrationTest/java/org/apache/ignite/internal/table/ItTableScanTest.java

```diff
@@ -590,14 +513,71 @@ public void testScanWithUpperBound() throws Exception {
                 null
         );
 
-        ArrayList<BinaryRow> scannedRows2 = scanAllRows(publisher2);
+        List<BinaryRow> scannedRows2 = scanAllRows(publisher2);
 
         assertEquals(5, scannedRows2.size());
 
         log.info("Result of scanning after insert rows: " + scannedRows2.stream().map(binaryRow -> rowToString(binaryRow))
                 .collect(Collectors.joining(", ")));
     }
 
+    @Test
+    public void testPhantomReads() throws Exception {
+        int iterations = 10;
+
+        // "for" is better at detecting data races than RepeatedTest.
+        for (int i = 0; i < iterations; i++) {
+            KeyValueView<Tuple, Tuple> kvView = table.keyValueView();
+
+            UUID sortedIndexId = getSortedIndexId();
+
+            int partId = 0;
+
+            InternalTransaction tx = startTxWithEnlistedPartition(partId);
+
+            try {
+                IgniteBiTuple<ClusterNode, Long> leaderWithTerm = tx.enlistedNodeAndTerm(new TablePartitionId(table.tableId(), partId));
+
+                PrimaryReplica recipient = new PrimaryReplica(leaderWithTerm.get1(), leaderWithTerm.get2());
+
+                Publisher<BinaryRow> publisher = internalTable.scan(partId, tx.id(), recipient, sortedIndexId, null, null, 0, null);
+
+                // Non-thread-safe collection is fine, HB is guaranteed by "Thread#join" inside of "runRace".
```

Comment:
How about just replacing the collection with a thread-safe one, so that the thread-safety question do not even arise? It's a test, performance does not seem to be critical here. Then the comment is not needed at all.

Reply:
I don't like using thread-safe collections when no one needs them, I consider this a bad practice.